### PR TITLE
Implement a harness for s_swap in priority_queue.c

### PR DESCRIPTION
**Pull request by angelhof**
Implement a proof harness for `s_swap` in `priority_queue.c`.

In the process I extended `priority_queue_is_valid` to check
whether the backpointers array_list is correctly allocated to be of equal
size to the container array_list.

Note that the `Makefile.common` is taken from #373 because `s_swap` is a static function.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_priority_queue_s_swap/aws_priority_queue_s_swap_harness.c [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/priority_queue.h>
+#include <proof_helpers/make_common_data_structures.h>
+
+void __CPROVER_file_local_priority_queue_c_s_swap(struct aws_priority_queue *queue, size_t a, size_t b);
+
+/**
+ * Runtime: 180s
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_priority_queue_s_swap/aws_priority_queue_s_swap_harness.c]: I've been putting these in the makefile, next to the size selected.  But either place works.

File: .cbmc-batch/jobs/aws_priority_queue_s_swap/aws_priority_queue_s_swap_harness.c [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    __CPROVER_assume(aws_priority_queue_is_bounded(&queue, MAX_INITIAL_ITEM_ALLOCATION, MAX_ITEM_SIZE));
+    ensure_priority_queue_has_allocated_members(&queue);
+
+    /* Assume the function preconditions */
+    __CPROVER_assume(aws_priority_queue_is_valid(&queue));
+
+    /* Perform operation under verification */
+    size_t a;
+    size_t b;
+    __CPROVER_assume(a < queue.container.length);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_priority_queue_s_swap/aws_priority_queue_s_swap_harness.c]: these are function preconditions?

File: source/priority_queue.c [hunk 3]
```diff
@@ -22,6 +22,8 @@
 #define RIGHT_OF(index) (((index) << 1) + 2)
 
 static void s_swap(struct aws_priority_queue *queue, size_t a, size_t b) {
+    AWS_PRECONDITION(aws_priority_queue_is_valid(queue));
+
```
Comment (danielsn) [inline source/priority_queue.c]: precondition that a and b are in range

File: source/priority_queue.c [hunk 4]
```diff
@@ -157,13 +160,31 @@ void aws_priority_queue_init_static(
...
+     * aws_priority_queue_nodes */
+    bool backpointer_list_item_size = queue->backpointers.item_size == sizeof(struct aws_priority_queue_node *);
+    bool lists_equal_lengths = queue->backpointers.length == queue->container.length;
+    bool backpointers_uninitialized =
+        (queue->backpointers.alloc == NULL && queue->backpointers.current_size == 0 &&
+         queue->backpointers.length == 0 && queue->backpointers.item_size == 0 && queue->backpointers.data == NULL);
+    bool backpointer_struct_is_valid =
+        backpointers_uninitialized || (backpointer_list_item_size && lists_equal_lengths);
+
+    /* bool backpointer_validity = s_backpointers_are_valid(&queue->backpointers); */
```
Comment (feliperodri) [inline source/priority_queue.c]: why is this commented out? @angelhof

Comment (angelhof) [reply source/priority_queue.c]: I forgot to remove it, thanks.

**Comment by angelhof** [comment]
Could you review and merge this PR @karkhaz ?

**Comment by angelhof** [comment]
@ColdenCullen @JonathanHenson @justinboswell @graebm @bretambrose could you review this change?

**Pull request merged**